SI-8667 Improve too-many-args message

som-snytt opened the pull request:
Use removeNames to help diagnose the application.
Supplement the error message with how many extra
args and any other residual assignments that the
user might have thought was a properly named arg.

som-snytt committed:
SI-8667 Improve too-many-args message

Use removeNames to help diagnose the application.
Supplement the error message with how many extra
args and any other residual assignments that the
user might have thought was a properly named arg.

The error message is gradual: succinct for short
arg lists, more verbose for longer applications.
Very long arg lists are probably generated, so
that message is the least colloquial.

janekdb commented:
This is an improvement but `25 too many arguments for method` could be regarded as the the more convoluted precursor to `25 more arguments than expected for method`.

Compare,

A | B
------------ | -------------
one too many arguments for method | one more argument than expected for method
2 too many arguments for method | 2 more arguments than expected for method
11 too many arguments for method | 11 more arguments than expected for method
25 too many arguments for method | 25 more arguments than expected for method

som-snytt commented:
I think "expected" isn't the right word, because of defaults and tupling.

`25 more arguments than can be applied to method f: (i: Int)Int`

janekdb commented:
That formulation looks good.

som-snytt commented:
I'll miss the colloquial terseness of "too many args", and why did it even spell out "arguments", but thanks for the nudge.

som-snytt commented:
I didn't squash, in case someone prefers this terser version:
```
t8667.scala:13: error: too many args (4) for constructor C: (a: Int, b: Int)C ; c, d are not parameters

//reduces counting but demands mental subtraction
t8667.scala:13: error: too many args (4 supplied but requires 2) for constructor C: (a: Int, b: Int)C ; c, d are not parameters

// instead of now
t8667.scala:13: error: 2 more arguments than can be applied to constructor C: (a: Int, b: Int)C ; c, d are not parameters
```

som-snytt commented:
I squashed with an update for gradual messages that depend on the arg count. Off by one for a short param list is different from off by 5 for a 22-arity generated function.

retronym commented:
Here's the current Javac version for comparison's sake:

```
 cat sandbox/Test.java  && javac sandbox/Test.java
class Test {
	void foo(int a) {};
	void bar() {
		foo(1, 2, 3);
	}
}
sandbox/Test.java:4: error: method foo in class Test cannot be applied to given types;
		foo(1, 2, 3);
		^
  required: int
  found: int,int,int
  reason: actual and formal argument lists differ in length
1 error
```

retronym commented:
An alternative (or additional ) idea: should be focus the error message at the first superfluous argument, rather than at the invoked method.

som-snytt commented:
I'd forgotten that missing-args was fiddled with a year ago: https://github.com/scala/scala/pull/4586

som-snytt commented:
Note to self, despite ambiguity of assignment, call out the erroneous `k=1`:
```
scala> def f(i: Int, j: Int) = i + j
f: (i: Int, j: Int)Int

scala> f(1,2,3)
<console>:13: error: too many arguments (3) for method f: (i: Int, j: Int)Int
       f(1,2,3)
             ^

scala> f(k = 1,i=2,j=3)
<console>:13: error: parameter 'i' is already specified at parameter position 1
Note that 'k' is not a parameter name of the invoked method.
       f(k = 1,i=2,j=3)
                ^
<console>:13: error: too many arguments (3) for method f: (i: Int, j: Int)Int
Note that 'k' is not a parameter name of the invoked method.
       f(k = 1,i=2,j=3)
                    ^
```

som-snytt committed:
SI-8667 Caret at bad arg

Pick the first excessive positional arg for the caret.

Note that erroring on named args doesn't do the obvious thing
in this regard.

If `k` was removed from the signature, then `f(k=1, i=2, j=3)`
doesn't tell us much about the wrong arg, because naming takes
the `k=1` as an assignment, `i` as duplicate naming. No arg is
deemed extra, though further inspection of the conflicting args
might get there. Since assignment syntax in parens is more|less
deprecated (?), no more effort is done here.

som-snytt commented:
My first job had me writing up the appendix of compiler error codes:

http://www.aztecmuseum.ca/docs/Aztec_C86_v3_2d_for_CPM86_Manual.pdf

I like that I wrote out "zero through seven", but other locutions are painfully awkward.

1: bad digit in octal constant
The only numerals permitted in the base 8 (octal) counting system are zero through
 seven.  In order to distinguish between octal, hexadecimal, and decimal constants, octal constants are preceded by a zero.  Any number beginning with a zero must not contain a digit greater than seven.  Octal constants look like this: 01, 027, 003.  Hexadecimal constants begin with 0x (e.g., 0x1, 0xAA0, 0xFFF).
2: string space exhausted
The compiler maintains an internal table of the strings appearing in the source code.  Since this table has a finite size, it may overflow during compilation and cause this error code.  The table default size is about one or two thousand characters depending on the operating system.  The size can be changed using the compiler option -Z.  Through simple guesswork, it is possible to arrive at a table size sufficient for compiling your program.
3: unterminated string
All strings must begin and end with double quotes (“).   This message indicates that a double quote has remained unpaired.
4: internal error
This error message should not occur.  It is a check on the internal workings of the compiler and is not known to be caused by any particular piece of code.  However, if this error code appears, please bring it to the attention of MANX.  It could be a bug in the compiler.  The release documentation enclosed with the product contains further information.
5:  illegal type for function
The type of a function refers to the type of the value which it returns.  Functions return an 
int by default unless they are declared otherwise.  However, functions are not allowed to return aggregates (arrays or structures).  An attempt to write a function such as  struct sam func() will generate this error code.  The legal function types are  char,  int, float, double,unsigned, long, void and a pointer to any type (including structures).
6:  error in argument declaration
The declaration list for the formal parameters of a function stands immediately before the left brace of the function body, as shown below.  Undeclared arguments default to int,  though it is usually better practice to declare everything. Naturally, this declaration list may be empty, whether or not the function takes any arguments at all.


11: data type too complex
This message is best explained by example:
  char *******foo;
The form of this declaration implies six pointers to pointers.   The seventh asterisk indicates a pointer to a char.   The compiler is unable to keep track of so many “levels”.  Removing just one of the asterisks will cure the error; all that is being declared in any case is a single two byte pointer. However it is to be hoped that such a construct will never be needed.

lrytz commented:
looks good to me - @janekdb @retronym can you take a look at https://github.com/scala/scala/pull/5147/files#diff-159fe691bd78013ef009e40722925f13

janekdb commented on test/files/neg/protected-constructors.check:
```diff
@@ -1,4 +1,4 @@
-protected-constructors.scala:17: error: too many arguments for constructor Foo1: ()dingus.Foo1
+protected-constructors.scala:17: error: no arguments allowed for nullary constructor Foo1: ()dingus.Foo1
```
+1

janekdb commented:
Exceeds my expectations so have a √LGTM leaving the deciding vote to @retronym.

som-snytt commented:
I haven't pushed retronym's idea to push the caret to the bad arg.

som-snytt commented:
Rebased. Moves the caret across the goal line.

som-snytt commented on test/files/neg/protected-constructors.check:
```diff
@@ -19,4 +19,7 @@ protected-constructors.scala:15: error: class Foo3 in object Ding cannot be acce
+protected-constructors.scala:15: error: no arguments allowed for nullary constructor Object: ()Object
+    class Bar3 extends Ding.Foo3("abc")
+                                 ^
+5 errors found
```
What's the fifth error?

som-snytt replied:
Did it start showing up because the position changed?

som-snytt replied:
Yes, in fact under `-Ydebug` it was:

```
test/files/neg/protected-constructors.scala:15: error: [ suppressed ] too many arguments for primary constructor Object: ()lang.this.Object
    class Bar3 extends Ding.Foo3("abc")
                            ^
```

som-snytt commented:
@retronym I moved the caret and a suppressed meaningless error is emitted. Is that a deal breaker? I didn't see a quick way to detect "I'd be suppressed at this other position."  Lo and behold, a decoupled interface with Reporter is actually inconvenient if I can't interrogate it. Or add API, `error(pos, msg)(suppressedByErrorAt = otherpos)`. Maybe the semantics is, "my underlying cause it at a different pos, at which suppression is computed".

adriaanm commented:
The Reporter API is certainly open to evolution (to simplify the process, we could have an internal subtype that has richer methods than the API used by sbt etc).

som-snytt commented:
Positions can be shifted, as in scripts that begin at an offset into the source file.

Analogously, you'd like to say, `error(pos withShiftedFocus displayFocus, "too bad")`. Then the error-suppressing reporter can check both the shifted focus and the underlying `pos`.

adriaanm commented:
LGTM

som-snytt commented:
I'll follow up in future about the spurious error. I didn't see where the inaccessible super ctor results in using nullary ctor. You'd hope the error wouldn't cascade. But I also like that last idea, where an error position (for display) has an underlying unshifted position (for semantics).

retronym commented:
> But I also like that last idea, where an error position (for display) has an underlying unshifted position (for semantics).

You could model something like that with the current API by issuing another error at the old position with a special message (e.g. empty string) that we could take as meaning it shouldn't be displayed.
